metricbeat/module/mongodb/dbstats/dbstats.go

```diff
@@ -56,47 +56,54 @@ func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
 	return &MetricSet{ms}, nil
 }
 
-// Fetch methods implements the data gathering and data conversion to the right format
-// It returns the event which is then forward to the output. In case of an error, a
-// descriptive error must be returned.
-func (m *MetricSet) Fetch() ([]common.MapStr, error) {
-	// events is the list of events collected from each of the databases.
-	var events []common.MapStr
-
+// Fetch methods implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *MetricSet) Fetch(reporter mb.ReporterV2) {
 	// instantiate direct connections to each of the configured Mongo hosts
 	mongoSession, err := mongodb.NewDirectSession(m.DialInfo)
 	if err != nil {
-		return nil, err
+		logger.Error(err)
+		reporter.Error(err)
+		return
 	}
 	defer mongoSession.Close()
 
 	// Get the list of databases names, which we'll use to call db.stats() on each
 	dbNames, err := mongoSession.DatabaseNames()
 	if err != nil {
-		logp.Err("Error retrieving database names from Mongo instance")
-		return events, err
+		err = errors.Wrap(err, "Error retrieving database names from Mongo instance")
+		logger.Error(err)
+		reporter.Error(err)
+		return
 	}
 
 	// for each database, call db.stats() and append to events
+	totalEvents := 0
 	for _, dbName := range dbNames {
 		db := mongoSession.DB(dbName)
 
 		result := common.MapStr{}
 
 		err := db.Run("dbStats", &result)
 		if err != nil {
-			logp.Err("Failed to retrieve stats for db %s", dbName)
+			err = errors.Wrapf(err, "Failed to retrieve stats for db %s", dbName)
+			logger.Error(err)
 			continue
 		}
 		data, _ := schema.Apply(result)
-		events = append(events, data)
+		if reported := reporter.Event(mb.Event{MetricSetFields: data}); !reported {
+			logger.Error("error reporting event")
```

Comment:
Notice that this is not an error, this is expected to happen when modules are stopped and we should not alert the user because of that.
In other places we are logging this at the debug level, and just in case we have to investigate something weird around this.